### PR TITLE
[libc++] Fix missing declarations of uses_allocator_construction_args

### LINKED ISSUE #66714: `uses_allocator_construction_args` has missing forward declarations
I was trying to use `std::uninitialized_construct_using_allocator` which was recently added in libc++, but one of the uses I had failed to compile. The actual problem is with `std::uses_allocator_construction_args`, so I'll put a simplified example using that instead.

```cpp
void test()
{
    using vt = std::pair<std::pmr::string, std::pair<std::pmr::string, std::pmr::string>>;

    std::pmr::polymorphic_allocator<vt> alloc;
    std::uses_allocator_construction_args<vt>(alloc, std::piecewise_construct,
                                              std::forward_as_tuple("foo"),
                                              std::forward_as_tuple("bar", "baz"));
}
```

This works with libstdc++ but fails to compile with libc++. After digging for a while, I found out the problem is that `__uses_allocator_construction_args` overloads are not forward-declared. In particular,

https://github.com/llvm/llvm-project/blob/c6472f571508113aceaa561e365cc0747b819a62/libcxx/include/__memory/uses_allocator_construction.h#L55-L57

is declared before

https://github.com/llvm/llvm-project/blob/c6472f571508113aceaa561e365cc0747b819a62/libcxx/include/__memory/uses_allocator_construction.h#L79-L81

which is why the instantiation of the first one fails (as it can't find the second declaration).

Godbolt repro: https://godbolt.org/z/WorEzejxW

I'd appreciate a fix. Thanks!

Note: this might be related to #64466

### PR DESCRIPTION
**Pull request by phyBrackets**
We were not declaring `__uses_allocator_construction_args` helper 
functions, leading to several valid uses failing to compile. This
patch solves the problem by moving these helper functions into a
struct, which also reduces the amount of redundant SFINAE we need
to perform since most overloads are checking for a cv-qualfied pair.

Fixes #66714

### CODE REVIEW

**Comment by phyBrackets** [comment]
Hi @philnik777 , thanks for the comment! I am kinda unsure about the current way of updating the tests is the best, have a look.

**Comment by phyBrackets** [comment]
not exactly sure about this, test seems to get passed on my machine.

**Comment by phyBrackets** [comment]
Not really sure why, test fail over modular build?

File: libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/uses_allocator_construction_args.pass.cpp
Comment (ldionne) [inline libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/uses_allocator_construction_args.pass.cpp:121]: You can change the test to this:

```
{
    // See https://github.com/llvm/llvm-project/issues/66714
    {
      using NestedPairType = std::pair<int, std::pair<int, UsesAllocArgT>>;
      std::same_as<std::tuple<std::piecewise_construct_t,
                              std::tuple<>,
                              std::tuple<std::piecewise_construct_t,
                                         std::tuple<>,
                                         std::tuple<std::allocator_arg_t, const Alloc&>>>> auto ret =
          test_uses_allocator_construction_args<NestedPairType>(a);
          (void)ret;
    }
    {
      using NestedPairType = std::pair<int, std::pair<UsesAllocArgT, int>>;
      std::same_as<std::tuple<std::piecewise_construct_t,
                              std::tuple<>,
                              std::tuple<std::piecewise_construct_t,
                                         std::tuple<std::allocator_arg_t, const Alloc&>,
                                         std::tuple<>>>> auto ret =
        test_uses_allocator_construction_args<NestedPairType>(a);
        (void)ret;
    }
  }
```

It doesn't require using `std::pmr`.

File: libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/common.h
Comment (ldionne) [inline libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/common.h:10]: Then the changes in this file can be removed.

**Review by ldionne — CHANGES_REQUESTED** [review]
I think this looks pretty good to me, but I am struggling to figure whether there's other similar cases that would be problematic and for which we should add test coverage.

@philnik777 can you take a look since you wrote this in the first place?

**Comment by ldionne** [comment]
Also, this should probably be rebased on top of #66939 once it lands -- I think it'll be easier to do that than the other way around.

**Comment by EricWF** [comment]
LGTM, but the bots need to get green and this change needs to be updated to merge cleanly.

**Comment by phyBrackets** [comment]
Thanks @ldionne and @EricWF !

File: libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/uses_allocator_construction_args.pass.cpp
Comment (philnik777) [inline libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/uses_allocator_construction_args.pass.cpp:125]: Could you add a short description instead of just a link to the bug report?

**Review by philnik777 — CHANGES_REQUESTED** [review]
This looks to me like we're adding a lot more forward declarations than tests, which makes me suspect that this is fixing bugs which we don't add regression tests for. Please make sure we test all the overloads (especially differently qualified pair overloads).

**Comment by philnik777** [comment]
It looks like you have some non-ASCII characters in your source. Did you copy the synopsis from eel.is/c++draft/? In that case they're probably in there.

**Comment by phyBrackets** [comment]
No, I don't think so 🤔. Although I have tried to look at any non-ASCII character in case, using regex but didn't find any.

**Comment by philnik777** [comment]
Oh sorry. I misread the output. It looks like the clang-format versions of the formatting checker and our generated output runner are different. Just add `libcxx/test/std/utilities/memory/allocator.uses/allocator.uses.construction/uses_allocator_construction_args.pass.cpp` to the `ignore_format.txt` and run the CI again.

**Comment by phyBrackets** [comment]
Thank you @philnik777 . Thoughts on the updated patch?

**Comment by philnik777** [comment]
Could you ping me again when the CI is green?

**Comment by ldionne** [comment]
It looks like this is only failing in C++20 mode now, hopefully this should be pretty easy to fix?

**Comment by ldionne** [comment]
Ping. I'd like to get this review finished and merged. If you have no time to work on it, please let us know and we'll find someone to pick it up, but we'd like to avoid accumulating too many outstanding reviews cause it makes the operation of the project harder.

**Comment by mordante** [comment]
@phyBrackets are you still interested in finishing this patch?

**Comment by ldionne** [comment]
I'll let @philnik777 finish the review on this one, don't let my "request for changes" block a merge.

**Comment by phyBrackets** [comment]
Hi @philnik777 , Can you please review this? I am not actually sure what causing CI to fails.

**Review by ldionne — APPROVED** [review]

**Pull request merged**
